drivers/sensor/bq2415x/bq2415x.c

```diff
@@ -0,0 +1,870 @@
+	case BQ2415X_PART_NUMBER:
+		return bq2415x_i2c_read_mask(bq2415x, BQ2415X_REG_VENDER, BQ2415X_MASK_PN,
+					     BQ2415X_SHIFT_PN);
+	case BQ2415X_REVISION:
+		return bq2415x_i2c_read_mask(bq2415x, BQ2415X_REG_VENDER,
+					     BQ2415X_MASK_REVISION,
+					     BQ2415X_SHIFT_REVISION);
+	}
+	return -EINVAL;
+}
+
+/* detect chip type */
+static enum bq2415x_chip bq2415x_detect_chip(struct bq2415x_device *bq2415x)
+{
+	const int ret = bq2415x_exec_command(bq2415x, BQ2415X_PART_NUMBER);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (ret) {
+	case 0:
+		if (bq2415x->chip == BQ24151A) {
+			return bq2415x->chip;
+		}
+		return BQ24151;
+	case 1:
+		if (bq2415x->chip == BQ24150A || bq2415x->chip == BQ24152 ||
+		    bq2415x->chip == BQ24155) {
+			return bq2415x->chip;
+		}
+		return BQ24150;
+	case 2:
+		if (bq2415x->chip == BQ24153A) {
+			return bq2415x->chip;
+		}
+		return BQ24153;
+	default:
+		return BQUNKNOWN;
+	}
+
+	return BQUNKNOWN;
+}
+
+/* detect chip revision */
+static int bq2415x_detect_revision(struct bq2415x_device *bq2415x)
+{
+	int ret = bq2415x_exec_command(bq2415x, BQ2415X_REVISION);
+	int chip = bq2415x_detect_chip(bq2415x);
+
+	if (ret < 0 || chip < 0) {
+		return -1;
+	}
+
+	switch (chip) {
+	case BQ24150:
+	case BQ24150A:
+	case BQ24151:
+	case BQ24151A:
+	case BQ24152:
+		if (ret >= 0 && ret <= 3) {
+			return ret;
+		}
+		return -1;
+	case BQ24153:
+	case BQ24153A:
+	case BQ24156:
+	case BQ24156A:
+	case BQ24157S:
+	case BQ24158:
+		if (ret == 3) {
+			return 0;
+		} else if (ret == 1) {
+			return 1;
+		}
+		return -1;
+	case BQ24155:
+		if (ret == 3) {
+			return 3;
+		}
+		return -1;
+	case BQUNKNOWN:
+		return -1;
+	}
+
+	return -1;
+}
+
+/* return chip vender code */
+static int bq2415x_get_vender_code(struct bq2415x_device *bq2415x)
+{
+	const int ret = bq2415x_exec_command(bq2415x, BQ2415X_VENDER_CODE);
+
+	if (ret < 0) {
+		return 0;
+	}
+
+	/* convert to binary */
+	return (ret & 0x1) + ((ret >> 1) & 0x1) * 10 + ((ret >> 2) & 0x1) * 100;
+}
+
+/* enable/disable auto resetting chip timer */
+static int bq2415x_set_autotimer(struct bq2415x_device *bq2415x, int state)
+{
+	if (state) {
+		k_delayed_work_submit(&bq2415x->timer_reset_work, K_SECONDS(BQ2415X_TIMER_TIMEOUT));
+		int ret;
+
+		ret = bq2415x_exec_command(bq2415x, BQ2415X_TIMER_RESET);
+		if (ret < 0) {
+			LOG_ERR("Failed to reset timer");
+			return ret;
+		}
+	} else {
+		k_delayed_work_cancel(&bq2415x->timer_reset_work);
+	}
+
+	return 0;
+}
+
+/* delayed work function for auto resetting chip timer */
+static void bq2415x_timer_work(struct k_work *work)
+{
+	struct bq2415x_device *bq2415x =
+		CONTAINER_OF(work, struct bq2415x_device, timer_reset_work);
+	int ret;
+
+	ret = bq2415x_exec_command(bq2415x, BQ2415X_TIMER_RESET);
+	if (ret < 0) {
+		LOG_ERR("Failed to reset timer.");
+		bq2415x_set_autotimer(bq2415x, 0);
+		return;
+	}
+	k_delayed_work_submit(&bq2415x->timer_reset_work, K_SECONDS(BQ2415X_TIMER_TIMEOUT));
```

Comment:
k_work API is changed, see https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#submitting-a-delayable-work-item

Reply:
Didn't know about it. Changed. Thanks!